### PR TITLE
Add comments

### PR DESCRIPTION
**Pull request by wolferbium**

### CODE REVIEW

File: spec/controllers/comments_controller_spec.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+require "spec_helper"
+
+RSpec.describe CommentsController, type: :controller do
+    describe "POST #create" do
+    context "with valid params" do
+      it "creates a new comment" do
+        post = create(:post)
+        expect {
+          create(:comment)
```
Comment (AlxGolubev) [inline spec/controllers/comments_controller_spec.rb]: It's not a test of a controller... It's a test for a factory. You need to make a request to controller, with params

And check that it creates instance, redirect's to where it should redirect, or render smth
Check that it has valid attributes after creation.

File: spec/controllers/posts_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,71 @@
+require "spec_helper"
+
+RSpec.describe PostsController, type: :controller do
+  describe "GET #index" do
+    it "has a 200 status code" do
+      post = create(:post)
```
Comment (AlxGolubev) [inline spec/controllers/posts_controller_spec.rb]: subject, let, before, it

File: spec/factories/user.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+FactoryBot.define do
+  factory :user do
+    email Faker::Internet.unique.email
+    password 'password'
+    password_confirmation 'password'
+  end
+
+  factory :admin, class: User do
```
Comment (AlxGolubev) [inline spec/factories/user.rb]: It can and even should be a trait, not a new factory

File: spec/models/user_spec.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+require 'spec_helper'
+
+RSpec.describe User, type: :model do
```
Comment (AlxGolubev) [inline spec/models/user_spec.rb]: ?

File: spec/controllers/comments_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+require "spec_helper"
```
Comment (AlxGolubev) [inline spec/controllers/comments_controller_spec.rb]: As I remember, it should be `rails_helper`

File: spec/controllers/posts_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,71 @@
+require "spec_helper"
```
Comment (AlxGolubev) [inline spec/controllers/posts_controller_spec.rb]: +

File: spec/controllers/users_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,31 @@
+require "spec_helper"
```
Comment (AlxGolubev) [inline spec/controllers/users_controller_spec.rb]: ++

File: spec/models/user_spec.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+require 'spec_helper'
```
Comment (AlxGolubev) [inline spec/models/user_spec.rb]: +

File: spec/spec_helper.rb [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+  config.order = "random"
+
+  config.include FactoryBot::Syntax::Methods
+
+  config.before(:suite) do
+    FactoryBot.find_definitions
+  end
+end
+
+Shoulda::Matchers.configure do |config|
```
Comment (AlxGolubev) [inline spec/spec_helper.rb]: Gem's specific configs are stored inside `support` library, and they have to be required using the following syntax

`Dir[Rails.root.join('spec/support/**/*.rb')].each { |f| require f }`

Comment (AlxGolubev) [reply spec/spec_helper.rb]: Actually, you already have `factory_bot` config inside `support` library

File: config/routes.rb [hunk 10]
```diff
@@ -1,9 +1,19 @@
...
   #mount Ckeditor::Engine => '/ckeditor'
-	root to: "posts#index"
+  root to: "posts#index"
+
+  resources :posts do
+    resources :comments, :users
+  end
+
+  resources :comments do
+    resources :comments
```
[the remark this answers is not in the document]
Comment (wolferbium) [reply config/routes.rb]: This line is needed to show linked comments.

File: Gemfile [hunk 11]
```diff
@@ -56,6 +56,16 @@ gem "haml-rails", "~> 1.0"
 
 gem "devise"
 
+group :development, :test do
+  gem 'rspec-rails', '~> 3.7'
+  gem 'factory_bot'
+  gem 'capybara'
+  gem 'faker'
+  gem 'rails-controller-testing'
```
[the remark this answers is not in the document]
Comment (wolferbium) [reply Gemfile]: rails-controller-testing позволяет использовать метод `assert_template` для проверки рендеринга страницы.

[the remark this answers is not in the document]
Comment (AlxGolubev) [reply Gemfile]: https://relishapp.com/rspec/rspec-rails/v/2-1/docs/controller-specs/render-views

File: app/assets/javascripts/comments.coffee [hunk 12]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (AlxGolubev) [inline app/assets/javascripts/comments.coffee]: do not store empty files in repo

File: app/controllers/comments_controller.rb [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+    if @comment.save(comment_params)
+      flash[:notice] = "Succesfully created post!"
+      redirect_to post_path(@commentable)
+    else
+      flash[:alert] = "Error creating new post!"
+      redirect_to post_path(@commentable)
+    end
+  end
+
+  def find_parent
```
Comment (AlxGolubev) [inline app/controllers/comments_controller.rb]: commentable

File: app/controllers/comments_controller.rb [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+    @commentable = Comment.find(params[:comment_id]) if params[:comment_id]
```
Comment (AlxGolubev) [inline app/controllers/comments_controller.rb]: use memoization

File: spec/controllers/comments_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,23 @@
+require "rails_helper"
+
+RSpec.describe CommentsController, type: :controller do
+ let(:user) { User.create email: "aTestUser@test.com", password: "aTestUserPsw", password_confirmation: "aTestUserPsw" }
```
Comment (AlxGolubev) [inline spec/controllers/comments_controller_spec.rb]: space

File: spec/controllers/comments_controller_spec.rb [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+  let(:other) { User.create email: "aTestUser2@test.com", password: "aTestUserPsw", password_confirmation: "aTestUserPsw" }
```
Comment (AlxGolubev) [inline spec/controllers/comments_controller_spec.rb]: variables naming

File: spec/controllers/comments_controller_spec.rb [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+  let!(:user_post) { Post.create title: "Hello", body: 'BodyOfThePost' }
+  let!(:user_post_comment) { Comment.create body: "Comment body to post", commentable: user_post, user: user }
+
+  context "#create" do
```
Comment (AlxGolubev) [inline spec/controllers/comments_controller_spec.rb]:
```ruby
let(:user) { create :user }

before { sign_in user }

describe '#create' do
  let(:parent) { create :parent }
  let(:params) { { modelname: attributes_for(:entry, parent: parent) } }
  
  subject { post :create, params: params }

  #happy path

  context 'when parameters are invalid' do
    #fails path  
  end

  context 'when blablabla' do
  end
end
```

File: spec/controllers/comments_controller_spec.rb [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+    before { sign_in user } 
+    it "should save comment to post with valid data" do
+      expect { post :create, params: { comment: { body: "BodyOfThePost" }, post_id: user_post.to_param } }.to change(Comment, :count).by(1)
```
Comment (AlxGolubev) [inline spec/controllers/comments_controller_spec.rb]: `expect { subject }.to change ...`

File: spec/controllers/posts_controller_spec.rb [hunk 19]
```diff
@@ -0,0 +1,60 @@
+require "rails_helper"
```
Comment (AlxGolubev) [inline spec/controllers/posts_controller_spec.rb]: same thing

File: .gitignore [hunk 20]
```diff
@@ -17,3 +17,4 @@
 /yarn-error.log
 
 .byebug_history
+.env
```
Comment (AlxGolubev) [inline .gitignore]: configure your editor to automatically add empty line to the end of a file on save

**Comment by AlxGolubev** [comment]
Please remove compiled assets from repo

**Pull request merged**